data-processing/common/commands/database.py

```diff
@@ -1,11 +1,39 @@
 from abc import ABC
 from argparse import ArgumentParser
-from typing import Any
+from enum import Enum
+import logging
+from typing import Any, Optional
 
 from common.commands.base import ArxivBatchCommand, I, R
 from common.models import setup_database_connections
 
 
+class OutputForm(Enum):
```

Comment:
Style nit pick: would it be possible to make this a `Literal` type instead of an Enum?

Reply:
Can you say a little more about this? 

It seems like if I were to use `Literal`, I would want to do something like:
```diff
diff --git a/data-processing/common/upload_entities.py b/data-processing/common/upload_entities.py
index 32d0e12..0403fce 100644
--- a/data-processing/common/upload_entities.py
+++ b/data-processing/common/upload_entities.py
@@ -1,10 +1,9 @@
 import dataclasses
-from enum import Enum
 import json
 import logging
 import os
 from typing import Any, Dict, List, Optional, Tuple
-# from typing_extensions import Literal
+from typing_extensions import Literal
 
 from peewee import fn
 
@@ -21,30 +20,28 @@ from common.types import (
     S2Id,
 )
 
-# OutputForm = Literal["db", "file"]
-
-class OutputForm(Enum):
-    DB = "db"
-    FILE = "file"
+DBOutputForm = Literal["db"]
+FileOutputForm = Literal["file"]
+OutputForm = Literal[DBOutputForm, FileOutputForm]
 
 
 class OutputDetails:
-    def __init__(self, output_forms: List[str], output_dir: Optional[str]):
+    def __init__(self, output_forms: List[OutputForm], output_dir: Optional[str]):
         OutputDetails.validate(output_forms=output_forms, output_dir=output_dir)
-        self.output_forms: List[OutputForm] = [OutputForm(output_form) for output_form in output_forms]
-        self.output_dir: Optional[str] = output_dir
+        self.output_forms = output_forms
+        self.output_dir = output_dir
 
     @staticmethod
-    def validate(output_forms: List[str], output_dir: Optional[str]) -> None:
+    def validate(output_forms: List[OutputForm], output_dir: Optional[str]) -> None:
         msg = "Please include either both 'file' as an output form and an output dir, or neither."
-        cond = (OutputForm.FILE.value in output_forms) == (output_dir is not None)
+        cond = (FileOutputForm in output_forms) == (output_dir is not None)
         assert cond, msg
 
     def can_save_to_db(self) -> bool:
-        return OutputForm.DB in self.output_forms
+        return DBOutputForm in self.output_forms
 
     def can_save_to_file(self) -> bool:
-        return OutputForm.FILE in self.output_forms
+        return FileOutputForm in self.output_forms
 
 
 def get_or_create_data_version(paper_id: str) -> int:
diff --git a/data-processing/scripts/run_pipeline.py b/data-processing/scripts/run_pipeline.py
index 2eb5a73..efdcdc9 100644
--- a/data-processing/scripts/run_pipeline.py
+++ b/data-processing/scripts/run_pipeline.py
@@ -27,7 +27,7 @@ from common.commands.store_pipeline_log import StorePipelineLog
 from common.commands.store_results import DEFAULT_S3_LOGS_BUCKET, StoreResults
 from common.fetch_arxiv import FetchFromArxivException
 from common.make_digest import make_paper_digest
-from common.upload_entities import OutputDetails, OutputForm
+from common.upload_entities import DBOutputForm, OutputDetails, OutputForm
 from common.types import PipelineDigest
 from entities.definitions.commands.detect_definitions import DetectDefinitions
 
@@ -342,10 +342,9 @@ if __name__ == "__main__":
 
     parser.add_argument(
         "--output-forms",
-        type=str,
+        type=OutputForm,
         nargs='+',
-        choices=[c.value for c in OutputForm],
-        default=[OutputForm.DB.value],
+        default=[DBOutputForm],
         help=(
             f"How we want the output to be stored. "
             + "If you include file, you must also provide a value for output-dir."
```
But that doesn't work.

The version I can get to work is something like this:
```diff
diff --git a/data-processing/common/upload_entities.py b/data-processing/common/upload_entities.py
index 32d0e12..e18bb49 100644
--- a/data-processing/common/upload_entities.py
+++ b/data-processing/common/upload_entities.py
@@ -1,10 +1,9 @@
 import dataclasses
-from enum import Enum
 import json
 import logging
 import os
 from typing import Any, Dict, List, Optional, Tuple
-# from typing_extensions import Literal
+from typing_extensions import Literal
 
 from peewee import fn
 
@@ -21,30 +20,26 @@ from common.types import (
     S2Id,
 )
 
-# OutputForm = Literal["db", "file"]
-
-class OutputForm(Enum):
-    DB = "db"
-    FILE = "file"
+OutputForm = Literal["db", "file"]
 
 
 class OutputDetails:
-    def __init__(self, output_forms: List[str], output_dir: Optional[str]):
+    def __init__(self, output_forms: List[OutputForm], output_dir: Optional[str]):
         OutputDetails.validate(output_forms=output_forms, output_dir=output_dir)
-        self.output_forms: List[OutputForm] = [OutputForm(output_form) for output_form in output_forms]
-        self.output_dir: Optional[str] = output_dir
+        self.output_forms = output_forms
+        self.output_dir = output_dir
 
     @staticmethod
-    def validate(output_forms: List[str], output_dir: Optional[str]) -> None:
+    def validate(output_forms: List[OutputForm], output_dir: Optional[str]) -> None:
         msg = "Please include either both 'file' as an output form and an output dir, or neither."
-        cond = (OutputForm.FILE.value in output_forms) == (output_dir is not None)
+        cond = ("file" in output_forms) == (output_dir is not None)
         assert cond, msg
 
     def can_save_to_db(self) -> bool:
-        return OutputForm.DB in self.output_forms
+        return "db" in self.output_forms
 
     def can_save_to_file(self) -> bool:
-        return OutputForm.FILE in self.output_forms
+        return "file" in self.output_forms
 
 
 def get_or_create_data_version(paper_id: str) -> int:
diff --git a/data-processing/scripts/run_pipeline.py b/data-processing/scripts/run_pipeline.py
index 2eb5a73..1cd28e9 100644
--- a/data-processing/scripts/run_pipeline.py
+++ b/data-processing/scripts/run_pipeline.py
@@ -27,7 +27,7 @@ from common.commands.store_pipeline_log import StorePipelineLog
 from common.commands.store_results import DEFAULT_S3_LOGS_BUCKET, StoreResults
 from common.fetch_arxiv import FetchFromArxivException
 from common.make_digest import make_paper_digest
-from common.upload_entities import OutputDetails, OutputForm
+from common.upload_entities import OutputDetails
 from common.types import PipelineDigest
 from entities.definitions.commands.detect_definitions import DetectDefinitions
 
@@ -344,8 +344,8 @@ if __name__ == "__main__":
         "--output-forms",
         type=str,
         nargs='+',
-        choices=[c.value for c in OutputForm],
-        default=[OutputForm.DB.value],
+        choices=["db", "file"],
+        default=["db"],
         help=(
             f"How we want the output to be stored. "
             + "If you include file, you must also provide a value for output-dir."
```
Which seems worse than the current Enum option, because I've got "db" and "file" lots of places (which feels like it would be more prone to typos coming about).

Note: this is the first time I've seen python `Literal`s, so I'd love to know if I'm not using them quite right, and there's a way of using them that does work better!

Reply:
The second implementation is just the one I was thinking of! Though I think you make a good point:

> Which seems worse than the current Enum option, because I've got "db" and "file" lots of places (which feels like it would be more prone to typos coming about).

When I have used mypy and pylance in VSCode, it has supported autocomplete of literal values, and statically checks in cases where the string provided for the type is not one that is listed in the Literal type, to help rule out typos. Though that said, I think I good argument in favor of sticking with enums is if this support doesn't seem to be consistently in tools, where I think we could rely on good typo-checking support for enums.